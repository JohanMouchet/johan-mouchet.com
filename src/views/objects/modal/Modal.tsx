import React, { useState } from "react";
import { createPortal } from "react-dom";
import FocusTrap from "focus-trap-react";
import cx from "classnames";
import { IconX } from "views/objects";
import "./Modal.scss";

type Props = {
  isOpen?: boolean;
  size?: "sm" | "md";
  trigerLabel: string;
  closeLabel?: string;
  className?: string | string[] | { [key: string]: boolean };
  children: React.ReactNode;
};

const Modal: React.FC<Props> = ({
  isOpen,
  size = "md",
  trigerLabel,
  closeLabel = "Close Modal",
  className,
  children,
}) => {
  const [open, setOpen] = useState(isOpen);

  const openModal = () => {
    setOpen(true);
    toggleScrollLock();
  };

  const closeModal = () => {
    setOpen(false);
    toggleScrollLock();
  };

  const handleEscKey = (event: React.KeyboardEvent) => {
    event.keyCode === 27 && closeModal();
  };

  const handleOutsideClick = (event: React.MouseEvent) => {
    event.target === event.currentTarget && closeModal();
  };

  const toggleScrollLock = () => {
    document.querySelector("html")?.classList.toggle("u-scroll-lock");
  };

  return (
    <>
      <button
        onClick={openModal}
        className={cx("o-modal__trigger", className)}
        type="button"
      >
        {trigerLabel}
      </button>

      {!open
        ? null
        : createPortal(
            <FocusTrap>
              <div
                role="dialog"
                tabIndex={-1}
                aria-modal="true"
                className="o-modal__cover u-animation u-animation--fast u-animation--fade-in"
                onClick={handleOutsideClick}
                onKeyDown={handleEscKey}
              >
                <div
                  className={cx(
                    "o-modal",
                    "container",
                    size && `container--${size}`
                  )}
                >
                  <button
                    className="o-modal__close-button"
                    onClick={closeModal}
                    aria-label={closeLabel}
                  >
                    <IconX />
                  </button>
                  <div className="o-modal__body">{children}</div>
                </div>
              </div>
            </FocusTrap>,
            document.body
          )}
    </>
  );
};

export default Modal;
