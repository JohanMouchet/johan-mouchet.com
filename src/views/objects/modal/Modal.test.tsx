import React from "react";
import { render } from "@testing-library/react";
import Modal from ".";

it("Renders <Modal /> without crashing", () => {
  expect(
    render(
      <Modal
        isOpen={false}
        size="sm"
        trigerLabel="Open Modal"
        closeLabel="Close Modal"
      >
        children
      </Modal>
    ).container.firstChild
  ).toMatchSnapshot();
});
