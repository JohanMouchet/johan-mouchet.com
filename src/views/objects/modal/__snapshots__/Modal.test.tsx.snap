// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders <Modal /> without crashing 1`] = `
<button
  class="o-modal__trigger"
  type="button"
>
  Open Modal
</button>
`;
